zen-oft-v2/contracts/ZenNativeOFT.sol

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.22;
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import { OFT } from "@layerzerolabs/oft-evm/contracts/OFT.sol";
+import { SendParam, OFTReceipt } from "@layerzerolabs/oft-evm/contracts/interfaces/IOFT.sol";
+import { MessagingParams, MessagingFee, MessagingReceipt } from "@layerzerolabs/lz-evm-protocol-v2/contracts/interfaces/ILayerZeroEndpointV2.sol";
+
+
+contract ZenNativeOFT is OFT, ReentrancyGuard {
+
+    uint256 immutable public CONVERSION_RATE;
+
+    error FailedUnwrap();
+    error NotEnoughMsgValue();
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        address _lzEndpoint,
+        address _delegate
+    ) OFT(_name, _symbol, _lzEndpoint, _delegate) Ownable(_delegate) {
+        CONVERSION_RATE = 10**(18-decimals());
+    }
+
+    function decimals() public view virtual override returns (uint8) {
+        return 18;
+    }
+
+    /**
+     * @dev Retrieves the shared decimals of the OFT.
+     * @return The shared decimals of the OFT.
+     *
+     * @dev Sets an implicit cap on the amount of tokens, over uint64.max() will need some sort of outbound cap / totalSupply cap
+     * Lowest common decimal denominator between chains.
+     * Defaults to 6 decimal places to provide up to 18,446,744,073,709.551615 units (max uint64).
+     * For tokens exceeding this totalSupply(), they will need to override the sharedDecimals function with something smaller.
+     * ie. 4 sharedDecimals would be 1,844,674,407,370,955.1615
+     */
+    function sharedDecimals() public view virtual override returns (uint8) {
+        return 18;
+    }
+
+    /**
+     * @dev Executes the send operation.
+     * @param _sendParam The parameters for the send operation.
+     * @param _fee The calculated fee for the send() operation.
+     *      - nativeFee: The native fee.
+     *      - lzTokenFee: The lzToken fee.
+     * @param _refundAddress The address to receive any excess funds.
+     * @return msgReceipt The receipt for the send operation.
+     * @return oftReceipt The OFT receipt information.
+     *
+     * @dev MessagingReceipt: LayerZero msg receipt
+     *  - guid: The unique identifier for the sent message.
+     *  - nonce: The nonce of the sent message.
+     *  - fee: The LayerZero fee incurred for the message.
+     */
+
+    function send(
+        SendParam calldata _sendParam,
+        MessagingFee calldata _fee,
+        address _refundAddress
+    ) external payable virtual override returns (MessagingReceipt memory msgReceipt, OFTReceipt memory oftReceipt) {

```

Comment:
just curious why we are overriding some of these functions if they are inherited from LayerZero's OFT contract

Reply:
In this case, it's because we invoke the custom `_lzSendWithAmount` instead of `_lsZend`
I can't simply override the `_lzSend`` because I had to add the parameter amount, that it was not explicitely inserted before. Same for `_payNative`, the original one doesn't check that `msg.value` is amount + fee because the original isn't developed for auto-wrapping native tokens.

Similar for `_credit`, it is overriden since it has to send native token instead of minting erc20 token

Reply:
makes sense. just curious why we need to pass the `amount` explicitly and not use `msg.sender`?